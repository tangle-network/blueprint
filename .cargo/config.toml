[target.aarch64-apple-darwin]
linker = "clang++"