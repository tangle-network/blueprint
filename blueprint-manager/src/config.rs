use std::path::PathBuf;
use structopt::StructOpt;

#[derive(Debug, StructOpt)]
#[structopt(
    name = "Blueprint Manager",
    about = "An program executor that connects to the Tangle network and runs protocols dynamically on the fly"
)]
pub struct BlueprintManagerConfig {
    /// The path to the gadget configuration file
    #[structopt(parse(from_os_str), short = "s", long = "gadget-config")]
    pub gadget_config: Option<PathBuf>,
    /// The path to the keystore
    #[structopt(short = "k", long = "keystore-uri")]
    pub keystore_uri: String,
    /// The verbosity level, can be used multiple times
    #[structopt(long, short = "v", parse(from_occurrences))]
    pub verbose: i32,
    /// Whether to use pretty logging
    #[structopt(long)]
    pub pretty: bool,
    /// An optional unique string identifier for the blueprint manager to differentiate between multiple
    /// running instances of a BlueprintManager (mostly for debugging purposes)
    #[structopt(long, short = "id")]
    pub instance_id: Option<String>,
    #[structopt(long, short = "t")]
    pub test_mode: bool,
}
