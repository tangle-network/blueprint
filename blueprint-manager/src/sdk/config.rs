use gadget_common::environments::GadgetEnvironment;
pub use gadget_io::KeystoreConfig;
use libp2p::Multiaddr;
use std::net::IpAddr;
use std::path::PathBuf;
use tangle_subxt::tangle_testnet_runtime::api::runtime_types::tangle_primitives::services::ServiceBlueprint;

#[derive(Debug)]
pub struct SingleGadgetConfig<KBE: Clone, Env: GadgetEnvironment> {
    pub base_path: PathBuf,
    pub keystore: KeystoreConfig,
    pub environment: Env,
    pub bind_ip: IpAddr,
    pub bind_port: u16,
    pub bootnodes: Vec<Multiaddr>,
    pub services: Vec<ServiceBlueprint>,
    pub n_protocols: usize,
    pub keystore_backend: KBE,
}
