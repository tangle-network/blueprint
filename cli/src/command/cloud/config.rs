//! Cloud provider configuration management.
//!
//! This module handles the configuration and authentication setup for various cloud providers.
//! It provides interactive setup flows, credential management, and persistent configuration storage.

use clap::ValueEnum;
use color_eyre::{Result, eyre::Context};
use dialoguer::{Input, Password, Select, theme::ColorfulTheme};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;

/// Supported cloud providers for Blueprint deployment.
#[derive(Debug, Clone, Copy, ValueEnum, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum CloudProvider {
    #[value(name = "aws")]
    AWS,
    #[value(name = "gcp")]
    GCP,
    #[value(name = "azure")]
    Azure,
    #[value(name = "digitalocean", alias = "do")]
    DigitalOcean,
    #[value(name = "vultr")]
    Vultr,
}

impl std::fmt::Display for CloudProvider {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::AWS => write!(f, "AWS"),
            Self::GCP => write!(f, "Google Cloud"),
            Self::Azure => write!(f, "Azure"),
            Self::DigitalOcean => write!(f, "DigitalOcean"),
            Self::Vultr => write!(f, "Vultr"),
        }
    }
}

/// Cloud configuration storage.
///
/// Persisted to ~/.config/tangle/cloud.json
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CloudConfig {
    /// The default provider to use when none is specified
    pub default_provider: Option<CloudProvider>,
    /// Per-provider configuration settings
    pub providers: HashMap<CloudProvider, ProviderSettings>,
}

/// Provider-specific configuration settings.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderSettings {
    /// Default region for deployments
    pub region: String,
    /// GCP project ID (only used for Google Cloud)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    /// Internal flag indicating if provider is fully configured
    #[serde(skip)]
    pub configured: bool,
}

impl CloudConfig {
    /// Load config from disk or create default
    pub fn load() -> Result<Self> {
        let path = Self::config_path()?;

        if path.exists() {
            let content = std::fs::read_to_string(&path).context("Failed to read cloud config")?;
            // Try to parse as JSON first, fall back to TOML for backwards compatibility
            serde_json::from_str(&content)
                .or_else(|_| toml::from_str(&content))
                .context("Failed to parse cloud config")
        } else {
            Ok(Self::default())
        }
    }

    /// Save config to disk
    pub fn save(&self) -> Result<()> {
        let path = Self::config_path()?;

        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent).context("Failed to create config directory")?;
        }

        // For now, serialize to JSON (since toml serialization is not straightforward in v0.9)
        let content = serde_json::to_string_pretty(self).context("Failed to serialize config")?;

        std::fs::write(&path, content).context("Failed to write cloud config")?;

        Ok(())
    }

    fn config_path() -> Result<PathBuf> {
        let config_dir = dirs::config_dir()
            .ok_or_else(|| color_eyre::eyre::eyre!("Could not find config directory"))?;
        Ok(config_dir.join("tangle").join("cloud.json"))
    }
}

impl Default for CloudConfig {
    fn default() -> Self {
        Self {
            default_provider: None,
            providers: HashMap::new(),
        }
    }
}

/// Configure a cloud provider with interactive setup.
///
/// This function guides the user through provider-specific authentication setup,
/// including credential configuration, region selection, and default settings.
///
/// # Arguments
///
/// * `provider` - The cloud provider to configure
/// * `region` - Optional region override (otherwise prompts user)
/// * `set_default` - Whether to set this as the default provider
///
/// # Errors
///
/// Returns an error if:
/// * Configuration directory cannot be created
/// * Credentials are invalid or cannot be saved
/// * Provider-specific CLI tools are not available
///
/// # Examples
///
/// ```no_run
/// # use cargo_tangle::command::cloud::{configure, CloudProvider};
/// # async fn example() -> color_eyre::Result<()> {
/// // Configure AWS as default provider
/// configure(CloudProvider::AWS, Some("us-east-1".to_string()), true).await?;
/// # Ok(())
/// # }
/// ```
pub async fn configure(
    provider: CloudProvider,
    region: Option<String>,
    set_default: bool,
) -> Result<()> {
    println!("🔧 Configuring {}...\n", provider);

    let mut config = CloudConfig::load()?;

    // Get or prompt for region
    let region = if let Some(r) = region {
        r
    } else {
        prompt_region(provider)?
    };

    // Provider-specific setup
    match provider {
        CloudProvider::AWS => configure_aws().await?,
        CloudProvider::GCP => configure_gcp().await?,
        CloudProvider::Azure => configure_azure().await?,
        CloudProvider::DigitalOcean => configure_digitalocean().await?,
        CloudProvider::Vultr => configure_vultr().await?,
    }

    // Save settings
    let mut settings = ProviderSettings {
        region,
        project_id: None,
        configured: true,
    };

    // GCP needs project ID
    if provider == CloudProvider::GCP {
        settings.project_id = Some(Input::new().with_prompt("GCP Project ID").interact()?);
    }

    config.providers.insert(provider, settings);

    if set_default || config.default_provider.is_none() {
        config.default_provider = Some(provider);
    }

    config.save()?;

    println!("\n✅ {} configured successfully!", provider);
    if config.default_provider == Some(provider) {
        println!("   Set as default provider");
    }

    Ok(())
}

/// Configure AWS credentials
async fn configure_aws() -> Result<()> {
    // Check for existing AWS CLI config
    let aws_config = dirs::home_dir()
        .map(|h| h.join(".aws").join("credentials"))
        .filter(|p| p.exists());

    if let Some(_) = aws_config {
        println!("✓ Found AWS credentials in ~/.aws/credentials");
        return Ok(());
    }

    // Check environment variables
    if std::env::var("AWS_ACCESS_KEY_ID").is_ok() {
        println!("✓ Found AWS credentials in environment");
        return Ok(());
    }

    // Prompt for credentials
    println!("No AWS credentials found. Please provide:");
    println!("(These will be stored in ~/.aws/credentials)");

    let access_key = Input::<String>::new()
        .with_prompt("AWS Access Key ID")
        .interact()?;

    let secret_key = Password::new()
        .with_prompt("AWS Secret Access Key")
        .interact()?;

    // Save to ~/.aws/credentials
    let aws_dir = dirs::home_dir()
        .ok_or_else(|| color_eyre::eyre::eyre!("Could not find home directory"))?
        .join(".aws");

    std::fs::create_dir_all(&aws_dir)?;

    let credentials = format!(
        "[default]\naws_access_key_id = {}\naws_secret_access_key = {}\n",
        access_key, secret_key
    );

    std::fs::write(aws_dir.join("credentials"), credentials)?;

    Ok(())
}

/// Configure GCP credentials  
async fn configure_gcp() -> Result<()> {
    // Check for gcloud CLI
    if std::process::Command::new("gcloud")
        .arg("--version")
        .output()
        .is_ok()
    {
        println!("✓ Found gcloud CLI");

        // Check if already authenticated
        let output = std::process::Command::new("gcloud")
            .args(&[
                "auth",
                "list",
                "--filter=status:ACTIVE",
                "--format=value(account)",
            ])
            .output()?;

        if !output.stdout.is_empty() {
            let account = String::from_utf8_lossy(&output.stdout);
            println!("✓ Authenticated as {}", account.trim());
            return Ok(());
        }

        // Run gcloud auth
        println!("Running gcloud auth login...");
        std::process::Command::new("gcloud")
            .args(&["auth", "application-default", "login"])
            .status()?;
    } else {
        println!("⚠️  gcloud CLI not found");
        println!("   Please install: https://cloud.google.com/sdk/docs/install");
        println!("   Or set GOOGLE_APPLICATION_CREDENTIALS to a service account key file");
    }

    Ok(())
}

/// Configure Azure credentials
async fn configure_azure() -> Result<()> {
    // Check for az CLI
    if std::process::Command::new("az")
        .arg("--version")
        .output()
        .is_ok()
    {
        println!("✓ Found Azure CLI");

        // Check if logged in
        let output = std::process::Command::new("az")
            .args(&["account", "show"])
            .output()?;

        if output.status.success() {
            println!("✓ Already logged in to Azure");
            return Ok(());
        }

        // Run az login
        println!("Running az login...");
        std::process::Command::new("az").arg("login").status()?;
    } else {
        println!("⚠️  Azure CLI not found");
        println!("   Please install: https://aka.ms/azure-cli");
    }

    Ok(())
}

/// Configure DigitalOcean credentials
async fn configure_digitalocean() -> Result<()> {
    if std::env::var("DIGITALOCEAN_TOKEN").is_ok() {
        println!("✓ Found DigitalOcean token in environment");
        return Ok(());
    }

    println!("Get your API token from: https://cloud.digitalocean.com/account/api/tokens");

    let token = Password::new()
        .with_prompt("DigitalOcean API Token")
        .interact()?;

    // Save to .env file
    let env_file = std::env::current_dir()?.join(".env");
    let mut content = if env_file.exists() {
        std::fs::read_to_string(&env_file)?
    } else {
        String::new()
    };

    if !content.contains("DIGITALOCEAN_TOKEN") {
        content.push_str(&format!("\nDIGITALOCEAN_TOKEN={}\n", token));
        std::fs::write(env_file, content)?;
        println!("✓ Saved to .env file");
    }

    Ok(())
}

/// Configure Vultr credentials
async fn configure_vultr() -> Result<()> {
    if std::env::var("VULTR_API_KEY").is_ok() {
        println!("✓ Found Vultr API key in environment");
        return Ok(());
    }

    println!("Get your API key from: https://my.vultr.com/settings/#settingsapi");

    let api_key = Password::new().with_prompt("Vultr API Key").interact()?;

    // Save to .env file
    let env_file = std::env::current_dir()?.join(".env");
    let mut content = if env_file.exists() {
        std::fs::read_to_string(&env_file)?
    } else {
        String::new()
    };

    if !content.contains("VULTR_API_KEY") {
        content.push_str(&format!("\nVULTR_API_KEY={}\n", api_key));
        std::fs::write(env_file, content)?;
        println!("✓ Saved to .env file");
    }

    Ok(())
}

/// Prompt for region selection
fn prompt_region(provider: CloudProvider) -> Result<String> {
    let regions = match provider {
        CloudProvider::AWS => vec![
            ("us-east-1", "US East (N. Virginia)"),
            ("us-west-2", "US West (Oregon)"),
            ("eu-west-1", "Europe (Ireland)"),
            ("ap-northeast-1", "Asia Pacific (Tokyo)"),
        ],
        CloudProvider::GCP => vec![
            ("us-central1", "US Central (Iowa)"),
            ("us-west1", "US West (Oregon)"),
            ("europe-west1", "Europe (Belgium)"),
            ("asia-northeast1", "Asia (Tokyo)"),
        ],
        CloudProvider::Azure => vec![
            ("eastus", "East US"),
            ("westus2", "West US 2"),
            ("northeurope", "North Europe"),
            ("japaneast", "Japan East"),
        ],
        CloudProvider::DigitalOcean => vec![
            ("nyc3", "New York 3"),
            ("sfo3", "San Francisco 3"),
            ("ams3", "Amsterdam 3"),
            ("sgp1", "Singapore 1"),
        ],
        CloudProvider::Vultr => vec![
            ("ewr", "New Jersey"),
            ("lax", "Los Angeles"),
            ("ams", "Amsterdam"),
            ("nrt", "Tokyo"),
        ],
    };

    let display_regions: Vec<String> = regions
        .iter()
        .map(|(code, name)| format!("{} ({})", name, code))
        .collect();

    let selection = Select::with_theme(&ColorfulTheme::default())
        .with_prompt("Select region")
        .items(&display_regions)
        .default(0)
        .interact()?;

    Ok(regions[selection].0.to_string())
}

/// List all configured cloud providers.
///
/// Displays a formatted list of all configured providers with their settings,
/// including region, default status, and project IDs where applicable.
///
/// # Errors
///
/// Returns an error if the configuration file cannot be read.
///
/// # Examples
///
/// ```bash
/// cargo tangle cloud list
/// ```
pub async fn list_providers() -> Result<()> {
    let config = CloudConfig::load()?;

    if config.providers.is_empty() {
        println!("No cloud providers configured.");
        println!("Run `cargo tangle cloud configure <provider>` to get started.");
        return Ok(());
    }

    println!("Configured providers:\n");

    for (provider, settings) in &config.providers {
        let default = if Some(*provider) == config.default_provider {
            " (default)"
        } else {
            ""
        };

        println!("  {} {}", provider, default);
        println!("    Region: {}", settings.region);
        if let Some(project) = &settings.project_id {
            println!("    Project: {}", project);
        }
        println!();
    }

    Ok(())
}
