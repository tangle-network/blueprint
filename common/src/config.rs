use crate::client::{create_client, JobsClient};
pub use crate::debug_logger::DebugLogger;
use crate::environments::GadgetEnvironment;
pub use crate::module::network::Network;
pub use crate::module::GadgetProtocol;
pub use crate::prometheus::PrometheusConfig;
use async_trait::async_trait;

#[async_trait]
pub trait ProtocolConfig<Env: GadgetEnvironment>
where
    Self: Sized,
{
    type Network: Network<Env>;
    type Protocol: GadgetProtocol<Env>;
    type ProtocolSpecificConfiguration: Clone + NetworkAndProtocolSetup<Env> + Sync;
    fn params(&self) -> &Self::ProtocolSpecificConfiguration;

    fn take_network(&mut self) -> Self::Network;
    fn take_protocol(&mut self) -> Self::Protocol;
    fn take_client(&mut self) -> <Env as GadgetEnvironment>::Client;
    fn prometheus_config(&self) -> PrometheusConfig;

    async fn build(&self) -> Result<Self, crate::Error> {
        let jobs_client = self.params().build_jobs_client().await?;
        let (network, protocol) = self
            .params()
            .build_network_and_protocol(jobs_client)
            .await?;
        let client = self.params().client();
        let params = self.params().clone();
        let tx_manager = self.tx_manager();
        let logger = self.logger();
        let prometheus_config = self.prometheus_config();

        Ok(Self::new(
            network,
            client,
            protocol,
            params,
            tx_manager,
            logger,
            prometheus_config,
        ))
    }

    fn new(
        network: <Self::ProtocolSpecificConfiguration as NetworkAndProtocolSetup<Env>>::Network,
        client: <Env as GadgetEnvironment>::Client,
        protocol: <<Self as ProtocolConfig<Env>>::ProtocolSpecificConfiguration as NetworkAndProtocolSetup<Env>>::Protocol,
        params: Self::ProtocolSpecificConfiguration,
        tx_manager: <Env as GadgetEnvironment>::TransactionManager,
        logger: DebugLogger,
        prometheus_config: PrometheusConfig,
    ) -> Self;

    fn tx_manager(&self) -> <Env as GadgetEnvironment>::TransactionManager {
        self.params().tx_manager()
    }
    fn logger(&self) -> DebugLogger {
        self.params().logger()
    }

    async fn run(self) -> Result<(), crate::Error> {
        crate::run_protocol(self).await
    }
}

#[async_trait]
pub trait NetworkAndProtocolSetup<Env: GadgetEnvironment> {
    type Network;
    type Protocol;

    async fn build_jobs_client(&self) -> Result<JobsClient<Env>, crate::Error> {
        create_client(self.client(), self.logger(), self.tx_manager()).await
    }

    async fn build_network_and_protocol(
        &self,
        jobs_client: JobsClient<Env>,
    ) -> Result<(Self::Network, Self::Protocol), crate::Error>;
    fn tx_manager(&self) -> <Env as GadgetEnvironment>::TransactionManager;
    fn logger(&self) -> DebugLogger;
    fn client(&self) -> <Env as GadgetEnvironment>::Client;
}
