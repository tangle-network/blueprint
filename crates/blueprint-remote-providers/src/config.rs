//! Cloud provider configuration types

use serde::{Deserialize, Serialize};

/// Cloud provider configuration for all supported providers
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CloudConfig {
    pub enabled: bool,
    pub aws: Option<AwsConfig>,
    pub gcp: Option<GcpConfig>,
    pub azure: Option<AzureConfig>,
    pub digital_ocean: Option<DigitalOceanConfig>,
    pub vultr: Option<VultrConfig>,
}

/// AWS configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AwsConfig {
    pub enabled: bool,
    pub region: String,
    pub access_key: String,
    pub secret_key: String,
    pub priority: Option<u8>,
}

/// GCP configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GcpConfig {
    pub enabled: bool,
    pub region: String,
    pub project_id: String,
    pub service_account_path: String,
    pub priority: Option<u8>,
}

/// Azure configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AzureConfig {
    pub enabled: bool,
    pub region: String,
    pub client_id: String,
    pub client_secret: String,
    pub tenant_id: String,
    pub priority: Option<u8>,
}

/// DigitalOcean configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigitalOceanConfig {
    pub enabled: bool,
    pub region: String,
    pub api_token: String,
    pub priority: Option<u8>,
}

/// Vultr configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VultrConfig {
    pub enabled: bool,
    pub region: String,
    pub api_key: String,
    pub priority: Option<u8>,
}

impl CloudConfig {
    /// Load configuration from environment variables
    pub fn from_env() -> Option<Self> {
        use std::env;

        let mut cloud_config = CloudConfig::default();
        let mut any_enabled = false;

        // AWS configuration
        if let (Ok(key), Ok(secret)) = (
            env::var("AWS_ACCESS_KEY_ID"),
            env::var("AWS_SECRET_ACCESS_KEY"),
        ) {
            cloud_config.aws = Some(AwsConfig {
                enabled: true,
                region: env::var("AWS_DEFAULT_REGION").unwrap_or_else(|_| "us-east-1".to_string()),
                access_key: key,
                secret_key: secret,
                priority: Some(10),
            });
            any_enabled = true;
        }

        // GCP configuration
        if let Ok(project_id) = env::var("GCP_PROJECT_ID") {
            let service_account_path = env::var("GOOGLE_APPLICATION_CREDENTIALS")
                .unwrap_or_else(|_| "/etc/gcp/service-account.json".to_string());
            cloud_config.gcp = Some(GcpConfig {
                enabled: true,
                region: env::var("GCP_DEFAULT_REGION")
                    .unwrap_or_else(|_| "us-central1".to_string()),
                project_id,
                service_account_path,
                priority: Some(8),
            });
            any_enabled = true;
        }

        // Azure configuration
        if let (Ok(client_id), Ok(client_secret), Ok(tenant_id)) = (
            env::var("AZURE_CLIENT_ID"),
            env::var("AZURE_CLIENT_SECRET"),
            env::var("AZURE_TENANT_ID"),
        ) {
            cloud_config.azure = Some(AzureConfig {
                enabled: true,
                region: env::var("AZURE_DEFAULT_REGION").unwrap_or_else(|_| "East US".to_string()),
                client_id,
                client_secret,
                tenant_id,
                priority: Some(7),
            });
            any_enabled = true;
        }

        // DigitalOcean configuration
        if let Ok(token) = env::var("DO_API_TOKEN") {
            cloud_config.digital_ocean = Some(DigitalOceanConfig {
                enabled: true,
                region: env::var("DO_DEFAULT_REGION").unwrap_or_else(|_| "nyc3".to_string()),
                api_token: token,
                priority: Some(5),
            });
            any_enabled = true;
        }

        // Vultr configuration
        if let Ok(key) = env::var("VULTR_API_KEY") {
            cloud_config.vultr = Some(VultrConfig {
                enabled: true,
                region: env::var("VULTR_DEFAULT_REGION").unwrap_or_else(|_| "ewr".to_string()),
                api_key: key,
                priority: Some(3),
            });
            any_enabled = true;
        }

        if any_enabled {
            cloud_config.enabled = true;
            Some(cloud_config)
        } else {
            None
        }
    }
}
