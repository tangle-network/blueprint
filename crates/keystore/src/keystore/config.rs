use crate::remote::RemoteConfig;
use std::path::PathBuf;

#[derive(Default)]
pub struct KeystoreConfig {
    pub(crate) in_memory: bool,
    #[cfg(feature = "std")]
    pub(crate) fs_root: Option<PathBuf>,
    #[cfg(any(
        feature = "aws-signer",
        feature = "gcp-signer",
        feature = "ledger-browser",
        feature = "ledger-node"
    ))]
    pub(crate) remote_configs: Vec<RemoteConfig>,
}

impl KeystoreConfig {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn in_memory(mut self, value: bool) -> Self {
        self.in_memory = value;
        self
    }

    #[cfg(feature = "std")]
    pub fn fs_root(mut self, path: PathBuf) -> Self {
        self.fs_root = Some(path);
        self
    }

    cfg_remote! {
        pub fn remote(mut self, remote_config: RemoteConfig) -> Self {
            self.remote_configs.push(remote_config);
            self
        }
    }

    fn is_empty(&self) -> bool {
        let mut is_empty = self.in_memory;
        #[cfg(feature = "std")]
        {
            is_empty |= self.fs_root.is_none();
        }
        #[cfg(any(
            feature = "aws-signer",
            feature = "gcp-signer",
            feature = "ledger-browser",
            feature = "ledger-node"
        ))]
        {
            is_empty |= !self.remote_configs.is_empty();
        }

        is_empty
    }

    pub(crate) fn finalize(self) -> Self {
        if self.is_empty() {
            return self.in_memory(true);
        }

        self
    }
}
