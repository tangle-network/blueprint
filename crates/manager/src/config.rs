use crate::error::{Error, Result};
use clap::Parser;
use docktopus::bollard::system::Version;
use docktopus::bollard::{API_DEFAULT_VERSION, Docker};
use http_body_util::Full;
use hyper::body::Bytes;
use hyper::header::HeaderValue;
use hyper_util::client::legacy::Client;
use hyper_util::rt::TokioExecutor;
use std::fmt::Display;
use std::net::{IpAddr, Ipv4Addr};
use std::path::PathBuf;
use std::sync::LazyLock;
use tracing::error;
use url::Url;

pub static DEFAULT_DOCKER_HOST: LazyLock<Url> =
    LazyLock::new(|| Url::parse("unix:///var/run/docker.sock").unwrap());

#[derive(Debug, Parser)]
#[command(
    name = "Blueprint Manager",
    about = "An program executor that connects to the Tangle network and runs blueprints dynamically on the fly"
)]
pub struct BlueprintManagerConfig {
    /// The path to the blueprint configuration file
    #[arg(short = 's', long)]
    pub blueprint_config: Option<PathBuf>,
    /// The path to the keystore
    #[arg(short = 'k', long)]
    pub keystore_uri: String,
    /// The directory in which all blueprints will store their data
    #[arg(long, short = 'd', default_value = "./data")]
    pub data_dir: PathBuf,
    /// The cache directory for blueprint manager downloads
    #[arg(long, short = 'd', default_value_os_t = default_cache_dir())]
    pub cache_dir: PathBuf,
    /// The verbosity level, can be used multiple times to increase verbosity
    #[arg(long, short = 'v', action = clap::ArgAction::Count)]
    pub verbose: u8,
    /// Whether to use pretty logging
    #[arg(long)]
    pub pretty: bool,
    /// An optional unique string identifier for the blueprint manager to differentiate between multiple
    /// running instances of a `BlueprintManager` (mostly for debugging purposes)
    #[arg(long, alias = "id")]
    pub instance_id: Option<String>,
    #[arg(long, short = 't')]
    pub test_mode: bool,
    /// Whether to allow invalid GitHub attestations (binary integrity checks)
    ///
    /// This will also allow for running the manager without the GitHub CLI installed.
    #[arg(long)]
    pub allow_unchecked_attestations: bool,
    /// The preferred way to run a blueprint.
    ///
    /// This is not a guarantee that the blueprint will use this method, as there may not be a source
    /// available of this type.
    #[arg(long, short, default_value_t)]
    pub preferred_source: SourceType,
    /// The location of the Podman-Docker socket
    #[arg(long, short, default_value_t = DEFAULT_DOCKER_HOST.clone())]
    pub podman_host: Url,
    /// Authentication proxy options
    #[command(flatten)]
    pub auth_proxy_opts: AuthProxyOpts,
}

fn default_cache_dir() -> PathBuf {
    match dirs::cache_dir() {
        Some(dir) => dir.join("blueprint-manager"),
        None => PathBuf::from("./blueprint-manager-cache"),
    }
}

impl Default for BlueprintManagerConfig {
    fn default() -> Self {
        Self {
            blueprint_config: None,
            keystore_uri: "./keystore".into(),
            data_dir: PathBuf::from("./data"),
            cache_dir: default_cache_dir(),
            verbose: 0,
            pretty: false,
            instance_id: None,
            test_mode: false,
            allow_unchecked_attestations: false,
            preferred_source: SourceType::default(),
            podman_host: DEFAULT_DOCKER_HOST.clone(),
            auth_proxy_opts: AuthProxyOpts::default(),
        }
    }
}

#[derive(clap::ValueEnum, Debug, Copy, Clone, Default, PartialEq, Eq)]
pub enum SourceType {
    Container,
    #[default]
    Native,
    Wasm,
}

impl Display for SourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SourceType::Container => write!(f, "container"),
            SourceType::Native => write!(f, "native"),
            SourceType::Wasm => write!(f, "wasm"),
        }
    }
}

pub struct SourceCandidates {
    pub container: Option<Url>,
    pub wasm_runtime: Option<String>,
    pub preferred_source: SourceType,
}

impl SourceCandidates {
    /// Determine all runtime sources available on this system
    ///
    /// # Errors
    ///
    /// If the `preferred_source` is not available on this system, or there is an error during its
    /// detection. Errors that occur while searching for non-preferred sources are silently ignored.
    pub async fn load(preferred_source: SourceType, podman_host: Url) -> Result<SourceCandidates> {
        let mut ret = SourceCandidates {
            container: None,
            wasm_runtime: None,
            preferred_source,
        };

        if let Err(e) = ret.determine_podman(podman_host).await {
            if preferred_source == SourceType::Container {
                error!("Podman not found, cannot use container source type as default: {e}");
                return Err(e);
            }
        }

        if let Err(e) = ret.determine_wasm().await {
            if preferred_source == SourceType::Wasm {
                error!("No WASM runtime found, cannot use WASM source type as default: {e}");
                return Err(e);
            }
        }

        Ok(ret)
    }

    async fn determine_podman(&mut self, host: Url) -> Result<()> {
        fn check_server_header(server: Option<&HeaderValue>) -> bool {
            if let Some(server) = server {
                if let Ok(server) = server.to_str() {
                    return server.to_lowercase().contains("libpod");
                }
            }
            false
        }

        let client = Docker::connect_with_local(host.as_str(), 20, API_DEFAULT_VERSION)
            .map_err(|e| Error::Other(e.to_string()))?;

        // Check the version, cheapest route
        let ver: Version = client
            .version()
            .await
            .map_err(|e| Error::Other(format!("Unable to determine the Podman version: {e}")))?;
        if let Some(comps) = &ver.components {
            if comps
                .iter()
                .any(|c| c.name.to_lowercase().contains("podman"))
            {
                self.container = Some(host);
                return Ok(());
            }
        }
        if let Some(platform) = &ver.platform {
            if platform.name.to_lowercase().contains("podman") {
                self.container = Some(host);
                return Ok(());
            }
        }

        // Fallback, read the HTTP "Server" header from a /_ping
        let res = if let Some(socket) = host.as_str().strip_prefix("unix://") {
            let client = Client::builder(TokioExecutor::new())
                .build::<_, Full<Bytes>>(hyperlocal::UnixConnector);
            client
                .get(hyperlocal::Uri::new(socket, "/_ping").into())
                .await
                .map_err(|e| Error::Other(format!("Unable to reach specified Podman host: {e}")))?
        } else {
            let client = Client::builder(TokioExecutor::new()).build_http::<Full<Bytes>>();
            client
                .get(
                    format!("{host}/_ping")
                        .parse()
                        .map_err(|e| Error::Other(format!("Unable to parse provided URI: {e}")))?,
                )
                .await
                .map_err(|e| Error::Other(format!("Unable to reach specified Podman host: {e}")))?
        };

        if check_server_header(res.headers().get("Server")) {
            self.container = Some(host);
            return Ok(());
        }

        Err(Error::Other(String::from("No Podman-Docker socket found")))
    }

    #[expect(clippy::unused_async, reason = "TBD")]
    async fn determine_wasm(&mut self) -> Result<bool> {
        // TODO: Verify WASM runtime installations
        Ok(true)
    }
}

/// The options for the auth proxy
#[derive(Debug, Parser, Clone)]
pub struct AuthProxyOpts {
    /// The host on which the auth proxy will listen
    #[arg(long, default_value = "0.0.0.0")]
    pub auth_proxy_host: IpAddr,
    /// The port on which the auth proxy will listen
    #[arg(long, default_value_t = 8276)]
    pub auth_proxy_port: u16,
}

impl Default for AuthProxyOpts {
    fn default() -> Self {
        Self {
            auth_proxy_host: IpAddr::V4(Ipv4Addr::UNSPECIFIED),
            auth_proxy_port: 8276, // T9 Mapping of TBPM (Tangle Blueprint Manager)
        }
    }
}
