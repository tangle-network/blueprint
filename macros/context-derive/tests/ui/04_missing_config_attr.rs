use gadget_sdk::config::StdGadgetConfiguration;
use gadget_sdk::ctx::KeystoreContext;

#[derive(KeystoreContext)]
struct MyContext {
    foo: String,
    sdk_config: StdGadgetConfiguration,
}

fn main() {}
