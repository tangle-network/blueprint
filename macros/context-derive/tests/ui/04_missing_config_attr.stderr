error: No field with #[config] attribute found, please add #[config] to the field that holds the `gadget_sdk::config::GadgetConfiguration`
 --> tests/ui/04_missing_config_attr.rs:5:8
  |
5 | struct MyContext {
  |        ^^^^^^^^^
